File: python/goondan/goondan/_schema.py
"""Interpreter for the closed keyword subset used by `goondan.schema.json`.

The TypeScript host runs the same interpreter over a byte-identical copy of the
schema, so both hosts must produce the same issue code, path, message and order
for the same document. Issues are carried as ``{code, segments, message}`` while
they are collected and are turned into ``{code, path, message}`` objects with an
RFC 6901 JSON Pointer when they are reported.
"""

from __future__ import annotations

import json
import math
import re
from functools import cmp_to_key
from importlib.resources import files
from typing import Any, Iterable, Sequence

Segment = str | int
Issue = dict[str, Any]

SUPPORTED_KEYWORDS = {
    "$schema", "$id", "$comment", "title", "description", "default", "$defs", "$ref",
    "type", "const", "enum", "properties", "required", "additionalProperties", "propertyNames",
    "minProperties", "items", "minItems", "uniqueItems", "minLength", "pattern", "exclusiveMinimum",
    "oneOf", "anyOf", "allOf", "not", "dependentSchemas",
}
_SCHEMA_MAPS = {"$defs", "properties", "dependentSchemas"}
_SCHEMA_ONE = {"additionalProperties", "propertyNames", "items", "not"}
_SCHEMA_LISTS = {"oneOf", "anyOf", "allOf"}
_SHAPE_CODES = {"schema.type", "schema.const", "schema.enum"}

_SCHEMA_CACHE: dict[str, Any] = {}


def schema_text() -> str:
    """The packaged copy of `spec/goondan.schema.json`."""
    if "text" not in _SCHEMA_CACHE:
        _SCHEMA_CACHE["text"] = files("goondan").joinpath("goondan.schema.json").read_text(encoding="utf-8")
    return _SCHEMA_CACHE["text"]


def schema() -> dict[str, Any]:
    if "schema" not in _SCHEMA_CACHE:
        _SCHEMA_CACHE["schema"] = json.loads(schema_text())
    return _SCHEMA_CACHE["schema"]


def unsupported_keywords(node: Any, at: str = "#") -> list[str]:
    """Keyword uses the interpreter does not implement, for the schema sync test."""
    if isinstance(node, bool):
        return []
    if not isinstance(node, dict):
        return [f"{at} is not a schema"]
    problems: list[str] = []
    for key, value in node.items():
        if key not in SUPPORTED_KEYWORDS:
            problems.append(f"{at}/{key}")
        elif key in _SCHEMA_MAPS:
            for name, child in value.items():
                problems.extend(unsupported_keywords(child, f"{at}/{key}/{name}"))
        elif key in _SCHEMA_ONE:
            problems.extend(unsupported_keywords(value, f"{at}/{key}"))
        elif key in _SCHEMA_LISTS:
            for index, child in enumerate(value):
                problems.extend(unsupported_keywords(child, f"{at}/{key}/{index}"))
        elif key == "$ref" and not str(value).startswith("#/"):
            problems.append(f"{at}/$ref is not local")
    return problems


def json_type(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, (int, float)):
        return "number" if math.isfinite(value) else "invalid"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "object"
    return "invalid"


def json_equal(left: Any, right: Any) -> bool:
    kind = json_type(left)
    if kind != json_type(right):
        return False
    if kind == "array":
        return len(left) == len(right) and all(json_equal(a, b) for a, b in zip(left, right))
    if kind == "object":
        return left.keys() == right.keys() and all(json_equal(left[key], right[key]) for key in left)
    return left == right


def pointer(segments: Sequence[Segment]) -> str:
    return "".join("/" + str(segment).replace("~", "~0").replace("/", "~1") for segment in segments)


def issue(code: str, segments: Sequence[Segment], message: str) -> Issue:
    return {"code": code, "segments": list(segments), "message": message}


def _format(value: Any) -> str:
    try:
        return json.dumps(value, ensure_ascii=False)
    except (TypeError, ValueError):
        return repr(value)


def _resolve_ref(root: Any, ref: str) -> Any:
    node = root
    for raw in ref[2:].split("/"):
        node = node.get(raw.replace("~1", "/").replace("~0", "~")) if isinstance(node, dict) else None
        if node is None:
            raise ValueError(f"unresolvable $ref {ref}")
    return node


def _required_only(branches: Sequence[Any]) -> list[str] | None:
    keys: list[str] = []
    for branch in branches:
        if not isinstance(branch, dict) or len(branch) != 1 or not isinstance(branch.get("required"), list) or len(branch["required"]) != 1:
            return None
        keys.append(branch["required"][0])
    return keys


def _describe(root: Any, branch: Any) -> str:
    node = branch
    while isinstance(node, dict) and "$ref" in node and len(node) == 1:
        node = _resolve_ref(root, node["$ref"])
    if not isinstance(node, dict):
        return "an allowed form"
    if "const" in node:
        return _format(node["const"])
    if "enum" in node:
        return " or ".join(_format(item) for item in node["enum"])
    if node.get("type") == "object" and isinstance(node.get("required"), list) and node["required"]:
        return "{" + ", ".join(node["required"]) + "}"
    if "type" in node:
        return " or ".join(node["type"]) if isinstance(node["type"], list) else node["type"]
    return "an allowed form"


def _unique(items: Iterable[str]) -> list[str]:
    return list(dict.fromkeys(items))


def _combine(root: Any, kind: str, branches: Sequence[Any], results: Sequence[list[Issue]], at: list[Segment]) -> list[Issue]:
    keys = _required_only(branches)
    if keys:
        return [issue(f"schema.{kind}", at, f"must declare {'at least' if kind == 'anyOf' else 'exactly'} one of {', '.join(keys)}")]
    here = pointer(at)
    compatible = [
        index for index, errors in enumerate(results)
        if not any(error["code"] in _SHAPE_CODES and pointer(error["segments"]) == here for error in errors)
    ]
    if len(compatible) == 1:
        return list(results[compatible[0]])
    return [issue(f"schema.{kind}", at, "must be " + " or ".join(_unique(_describe(root, branch) for branch in branches)))]


def _check(root: Any, node: Any, value: Any, at: list[Segment]) -> list[Issue]:
    if node is True:
        return []
    if node is False:
        return [issue("schema.false", at, "is not allowed")]
    out: list[Issue] = []
    if "$ref" in node:
        out.extend(_check(root, _resolve_ref(root, node["$ref"]), value, at))
    kind = json_type(value)
    if "type" in node:
        types = node["type"] if isinstance(node["type"], list) else [node["type"]]
        if not any(expected == kind or (expected == "integer" and kind == "number" and float(value).is_integer()) for expected in types):
            out.append(issue("schema.type", at, "must be " + " or ".join(types)))
            return out
    if "const" in node and not json_equal(value, node["const"]):
        out.append(issue("schema.const", at, f"must be {_format(node['const'])}"))
    if "enum" in node and not any(json_equal(value, candidate) for candidate in node["enum"]):
        out.append(issue("schema.enum", at, "must be one of " + ", ".join(_format(item) for item in node["enum"])))
    if kind == "string" and "minLength" in node and len(value) < node["minLength"]:
        out.append(issue("schema.minLength", at, "must not be empty" if node["minLength"] == 1 else f"must contain at least {node['minLength']} characters"))
    if kind == "string" and "pattern" in node and re.search(node["pattern"], value) is None:
        out.append(issue("schema.pattern", at, f"must match {node['pattern']}"))
    if kind == "number" and "exclusiveMinimum" in node and not value > node["exclusiveMinimum"]:
        out.append(issue("schema.exclusiveMinimum", at, f"must be greater than {_format(node['exclusiveMinimum'])}"))
    if kind == "array":
        if "minItems" in node and len(value) < node["minItems"]:
            out.append(issue("schema.minItems", at, "must not be empty" if node["minItems"] == 1 else f"must contain at least {node['minItems']} items"))
        if node.get("uniqueItems") is True:
            for index, item in enumerate(value):
                if any(json_equal(earlier, item) for earlier in value[:index]):
                    out.append(issue("schema.uniqueItems", [*at, index], "duplicates an earlier item"))
        if "items" in node:
            for index, item in enumerate(value):
                out.extend(_check(root, node["items"], item, [*at, index]))
    if kind == "object":
        keys = list(value.keys())
        if "minProperties" in node and len(keys) < node["minProperties"]:
            out.append(issue("schema.minProperties", at, "must not be empty" if node["minProperties"] == 1 else f"must contain at least {node['minProperties']} entries"))
        for name in node.get("required", []):
            if name not in value:
                out.append(issue("schema.required", [*at, name], "is required"))
        for key in keys:
            if "propertyNames" in node:
                bad = _check(root, node["propertyNames"], key, [*at, key])
                if bad:
                    out.extend(issue("schema.propertyNames", [*at, key], error["message"]) for error in bad)
                    continue
            if key in node.get("properties", {}):
                out.extend(_check(root, node["properties"][key], value[key], [*at, key]))
            elif node.get("additionalProperties") is False:
                out.append(issue("schema.additionalProperties", [*at, key], "is not a supported field"))
            elif "additionalProperties" in node:
                out.extend(_check(root, node["additionalProperties"], value[key], [*at, key]))
        for name, dependent in node.get("dependentSchemas", {}).items():
            if name in value:
                out.extend(_check(root, dependent, value, at))
    for branch in node.get("allOf", []):
        out.extend(_check(root, branch, value, at))
    if "anyOf" in node:
        results = [_check(root, branch, value, at) for branch in node["anyOf"]]
        if not any(not errors for errors in results):
            out.extend(_combine(root, "anyOf", node["anyOf"], results, at))
    if "oneOf" in node:
        results = [_check(root, branch, value, at) for branch in node["oneOf"]]
        passing = sum(1 for errors in results if not errors)
        if passing == 0:
            out.extend(_combine(root, "oneOf", node["oneOf"], results, at))
        elif passing > 1:
            keys = _required_only(node["oneOf"])
            out.append(issue("schema.oneOf", at, f"must declare exactly one of {', '.join(keys)}" if keys else "matches more than one allowed form"))
    if "not" in node and not _check(root, node["not"], value, at):
        negated = node["not"]
        out.append(issue("schema.not", at, f"must not be {_format(negated['const'])}" if isinstance(negated, dict) and "const" in negated else "uses a form that is not allowed"))
    return out


def validate(value: Any, at: Sequence[Segment] = ()) -> list[Issue]:
    """Check a composed configuration document against the whole schema."""
    root = schema()
    return _check(root, root, value, list(at))


def validate_definition(name: str, value: Any, at: Sequence[Segment] = ()) -> list[Issue]:
    """Check one value against a `$defs` entry, as the read phase does for extends and resources."""
    root = schema()
    return _check(root, root["$defs"][name], value, list(at))


def json_issues(value: Any, at: Sequence[Segment] = ()) -> list[Issue]:
    """Values that a composed configuration document may not contain."""
    out: list[Issue] = []
    segments = list(at)
    if isinstance(value, bool) or value is None or isinstance(value, str):
        return out
    if isinstance(value, (int, float)):
        if not math.isfinite(value):
            out.append(issue("config.not_json", segments, "must be a finite number"))
        return out
    if isinstance(value, list):
        for index, item in enumerate(value):
            out.extend(json_issues(item, [*segments, index]))
        return out
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                out.append(issue("config.not_json", segments, "must use string keys"))
                continue
            out.extend(json_issues(item, [*segments, key]))
        return out
    out.append(issue("config.not_json", segments, f"must be a JSON value, not {type(value).__name__}"))
    return out


def _compare_segments(left: Sequence[Segment], right: Sequence[Segment]) -> int:
    for a, b in zip(left, right):
        if isinstance(a, int) and isinstance(b, int):
            if a != b:
                return -1 if a < b else 1
            continue
        first, second = str(a), str(b)
        if first != second:
            return -1 if first < second else 1
    return len(left) - len(right)


def _compare_issues(left: Issue, right: Issue) -> int:
    """§구성 오류: `path` segment by segment, then `code`. `message` orders nothing."""
    delta = _compare_segments(left["segments"], right["segments"])
    if delta:
        return delta
    if left["code"] != right["code"]:
        return -1 if left["code"] < right["code"] else 1
    return 0


def report(issues: Iterable[Issue]) -> list[dict[str, str]]:
    """§구성 오류: de-duplicate by `path` and `code`, then sort, and render `{code, path, message}`.

    The item a stage produced first survives a duplicate, and `message` is used neither for
    the key nor for the order, so both hosts report the same items in the same order even
    when they word a message differently.
    """
    seen: set[tuple[str, str]] = set()
    kept: list[Issue] = []
    for item in issues:
        key = (pointer(item["segments"]), item["code"])
        if key in seen:
            continue
        seen.add(key)
        kept.append(item)
    return [
        {"code": item["code"], "path": pointer(item["segments"]), "message": item["message"]}
        for item in sorted(kept, key=cmp_to_key(_compare_issues))
    ]


def prefix(issues: Iterable[Issue], at: Sequence[Segment]) -> list[Issue]:
    """Move issues under another document position, as nested configurations need."""
    head = list(at)
    return [issue(item["code"], [*head, *item["segments"]], item["message"]) for item in issues]

File: packages/core/src/schema.ts
import schemaDocument from "./goondan.schema.json" with { type: "json" };
import { isJsonObject, isRecord, jsonEqual, jsonType, ownKeys, pointer, sortIssues, toJson, type PointerSegment } from "./json.ts";
import { type ConfigIssue, type ConfigIssueCode, type Json } from "./types.ts";

/**
 * The closed keyword set `spec/goondan.schema.json` may use. A schema edit that adds any other
 * keyword fails `test/schema.test.ts` instead of silently going unenforced.
 */
export const supportedSchemaKeywords: ReadonlySet<string> = new Set([
  "$schema", "$id", "$comment", "title", "description", "default", "$defs", "$ref",
  "type", "const", "enum", "properties", "required", "additionalProperties", "propertyNames",
  "minProperties", "items", "minItems", "uniqueItems", "minLength", "pattern", "exclusiveMinimum",
  "oneOf", "anyOf", "allOf", "not", "dependentSchemas",
]);

const schemaMapKeywords = new Set(["$defs", "properties", "dependentSchemas"]);
const schemaKeywords = new Set(["additionalProperties", "propertyNames", "items", "not"]);
const schemaListKeywords = new Set(["oneOf", "anyOf", "allOf"]);
const shapeCodes = new Set(["schema.type", "schema.const", "schema.enum"]);

type Schema = boolean | Record<string, Json>;

interface RawIssue { code: ConfigIssueCode; segments: PointerSegment[]; message: string }

function isSchema(value: unknown): value is Schema {
  return typeof value === "boolean" || isRecord(value);
}

function asSchema(value: unknown): Schema {
  if (!isSchema(value)) throw new Error("Schema nodes must be objects or booleans");
  return value;
}

function keyword(schema: Schema, name: string): Json | undefined {
  if (typeof schema === "boolean") return undefined;
  return schema[name];
}

function stringList(value: Json | undefined): string[] | undefined {
  if (typeof value === "string") return [value];
  if (Array.isArray(value) && value.every((item): item is string => typeof item === "string")) return value;
  return undefined;
}

function jsonArray(value: Json | undefined): Json[] | undefined {
  return Array.isArray(value) ? value : undefined;
}

const format = (value: Json | undefined): string => JSON.stringify(value ?? null) ?? "null";

function resolveRef(root: Schema, ref: string): Schema {
  if (!ref.startsWith("#/")) throw new Error(`Unsupported $ref ${ref}`);
  let node: Json | undefined = typeof root === "boolean" ? undefined : root;
  for (const raw of ref.slice(2).split("/")) {
    const key = raw.replaceAll("~1", "/").replaceAll("~0", "~");
    if (!isRecord(node)) throw new Error(`Unresolvable $ref ${ref}`);
    node = node[key];
  }
  if (node === undefined) throw new Error(`Unresolvable $ref ${ref}`);
  return asSchema(node);
}

/** The keys a combinator requires when every branch only asks for one property to be present. */
function requiredOnly(branches: readonly Json[]): string[] | undefined {
  const keys: string[] = [];
  for (const branch of branches) {
    if (!isRecord(branch)) return undefined;
    if (ownKeys(branch).length !== 1) return undefined;
    const required = jsonArray(branch.required);
    if (!required || required.length !== 1) return undefined;
    const name = required[0];
    if (typeof name !== "string") return undefined;
    keys.push(name);
  }
  return keys;
}

function describe(root: Schema, branch: Json): string {
  let node: Json = branch;
  while (isRecord(node) && typeof node.$ref === "string" && ownKeys(node).length === 1) {
    const resolved = resolveRef(root, node.$ref);
    if (typeof resolved === "boolean") return "an allowed form";
    node = resolved;
  }
  if (!isRecord(node)) return "an allowed form";
  if (Object.hasOwn(node, "const")) return format(node.const);
  const enumeration = jsonArray(node.enum);
  if (enumeration) return enumeration.map(format).join(" or ");
  const required = jsonArray(node.required);
  if (node.type === "object" && required && required.length > 0) return `{${required.map(String).join(", ")}}`;
  const types = stringList(node.type);
  if (types) return types.join(" or ");
  return "an allowed form";
}

const unique = (items: readonly string[]): string[] => items.filter((item, index) => items.indexOf(item) === index);

function combine(root: Schema, kind: "oneOf" | "anyOf", branches: readonly Json[], results: readonly RawIssue[][], at: PointerSegment[]): RawIssue[] {
  const keys = requiredOnly(branches);
  if (keys) return [{ code: `schema.${kind}`, segments: at, message: `must declare ${kind === "anyOf" ? "at least" : "exactly"} one of ${keys.join(", ")}` }];
  const here = pointer(at);
  const compatible: number[] = [];
  results.forEach((issues, index) => {
    if (!issues.some((issue) => shapeCodes.has(issue.code) && pointer(issue.segments) === here)) compatible.push(index);
  });
  const only = compatible[0];
  if (compatible.length === 1 && only !== undefined) {
    const chosen = results[only];
    if (chosen) return chosen;
  }
  if (compatible.length > 1) {
    const fewest = Math.min(...compatible.map((index) => results[index]?.length ?? Number.POSITIVE_INFINITY));
    const closest = compatible.filter((index) => results[index]?.length === fewest);
    const closestIndex = closest[0];
    if (closest.length === 1 && closestIndex !== undefined) {
      const chosen = results[closestIndex];
      if (chosen) return chosen;
    }
  }
  return [{ code: `schema.${kind}`, segments: at, message: `must be ${unique(branches.map((branch) => describe(root, branch))).join(" or ")}` }];
}

function check(root: Schema, schema: Schema, value: unknown, at: PointerSegment[]): RawIssue[] {
  if (schema === true) return [];
  if (schema === false) return [{ code: "schema.false", segments: at, message: "is not allowed" }];
  const out: RawIssue[] = [];
  const ref = keyword(schema, "$ref");
  if (typeof ref === "string") out.push(...check(root, resolveRef(root, ref), value, at));
  const type = jsonType(value);
  const expectedTypes = stringList(keyword(schema, "type"));
  if (expectedTypes) {
    const matched = expectedTypes.some((expected) =>
      expected === type || (expected === "integer" && type === "number" && typeof value === "number" && Number.isInteger(value)));
    if (!matched) {
      out.push({ code: "schema.type", segments: at, message: `must be ${expectedTypes.join(" or ")}` });
      return out;
    }
  }
  if (typeof schema !== "boolean" && Object.hasOwn(schema, "const") && !jsonEqual(value, schema.const)) {
    out.push({ code: "schema.const", segments: at, message: `must be ${format(schema.const)}` });
  }
  const enumeration = jsonArray(keyword(schema, "enum"));
  if (enumeration && !enumeration.some((candidate) => jsonEqual(value, candidate))) {
    out.push({ code: "schema.enum", segments: at, message: `must be one of ${enumeration.map(format).join(", ")}` });
  }
  const minLength = keyword(schema, "minLength");
  if (type === "string" && typeof value === "string" && typeof minLength === "number" && [...value].length < minLength) {
    out.push({ code: "schema.minLength", segments: at, message: minLength === 1 ? "must not be empty" : `must contain at least ${String(minLength)} characters` });
  }
  const pattern = keyword(schema, "pattern");
  if (type === "string" && typeof value === "string" && typeof pattern === "string" && !new RegExp(pattern, "u").test(value)) {
    out.push({ code: "schema.pattern", segments: at, message: `must match ${pattern}` });
  }
  const exclusiveMinimum = keyword(schema, "exclusiveMinimum");
  if (type === "number" && typeof value === "number" && typeof exclusiveMinimum === "number" && !(value > exclusiveMinimum)) {
    out.push({ code: "schema.exclusiveMinimum", segments: at, message: `must be greater than ${format(exclusiveMinimum)}` });
  }
  if (type === "array" && Array.isArray(value)) {
    const minItems = keyword(schema, "minItems");
    if (typeof minItems === "number" && value.length < minItems) {
      out.push({ code: "schema.minItems", segments: at, message: minItems === 1 ? "must not be empty" : `must contain at least ${String(minItems)} items` });
    }
    if (keyword(schema, "uniqueItems") === true) {
      value.forEach((item, index) => {
        if (value.slice(0, index).some((earlier) => jsonEqual(earlier, item))) {
          out.push({ code: "schema.uniqueItems", segments: [...at, index], message: "duplicates an earlier item" });
        }
      });
    }
    const items = keyword(schema, "items");
    if (items !== undefined) {
      const itemSchema = asSchema(items);
      value.forEach((item, index) => out.push(...check(root, itemSchema, item, [...at, index])));
    }
  }
  if (type === "object" && isRecord(value)) {
    const keys = ownKeys(value);
    const minProperties = keyword(schema, "minProperties");
    if (typeof minProperties === "number" && keys.length < minProperties) {
      out.push({ code: "schema.minProperties", segments: at, message: minProperties === 1 ? "must not be empty" : `must contain at least ${String(minProperties)} entries` });
    }
    for (const name of stringList(keyword(schema, "required")) ?? []) {
      if (!keys.includes(name)) out.push({ code: "schema.required", segments: [...at, name], message: "is required" });
    }
    const propertyNames = keyword(schema, "propertyNames");
    const properties = keyword(schema, "properties");
    const additional = keyword(schema, "additionalProperties");
    for (const key of keys) {
      if (propertyNames !== undefined) {
        const broken = check(root, asSchema(propertyNames), key, [...at, key]);
        if (broken.length > 0) {
          out.push(...broken.map((broken): RawIssue => ({ code: "schema.propertyNames", segments: [...at, key], message: broken.message })));
          continue;
        }
      }
      if (isRecord(properties) && Object.hasOwn(properties, key)) {
        out.push(...check(root, asSchema(properties[key]), value[key], [...at, key]));
      } else if (additional === false) {
        out.push({ code: "schema.additionalProperties", segments: [...at, key], message: "is not a supported field" });
      } else if (additional !== undefined) {
        out.push(...check(root, asSchema(additional), value[key], [...at, key]));
      }
    }
    const dependent = keyword(schema, "dependentSchemas");
    if (isRecord(dependent)) {
      for (const name of ownKeys(dependent)) {
        if (keys.includes(name)) out.push(...check(root, asSchema(dependent[name]), value, at));
      }
    }
  }
  for (const branch of jsonArray(keyword(schema, "allOf")) ?? []) out.push(...check(root, asSchema(branch), value, at));
  const anyOf = jsonArray(keyword(schema, "anyOf"));
  if (anyOf) {
    const results = anyOf.map((branch) => check(root, asSchema(branch), value, at));
    if (!results.some((issues) => issues.length === 0)) out.push(...combine(root, "anyOf", anyOf, results, at));
  }
  const oneOf = jsonArray(keyword(schema, "oneOf"));
  if (oneOf) {
    const results = oneOf.map((branch) => check(root, asSchema(branch), value, at));
    const passing = results.filter((issues) => issues.length === 0).length;
    if (passing === 0) out.push(...combine(root, "oneOf", oneOf, results, at));
    else if (passing > 1) {
      const keys = requiredOnly(oneOf);
      out.push({ code: "schema.oneOf", segments: at, message: keys ? `must declare exactly one of ${keys.join(", ")}` : "matches more than one allowed form" });
    }
  }
  const not = keyword(schema, "not");
  if (not !== undefined && check(root, asSchema(not), value, at).length === 0) {
    const negated = isRecord(not) && Object.hasOwn(not, "const") ? `must not be ${format(not.const)}` : "uses a form that is not allowed";
    out.push({ code: "schema.not", segments: at, message: negated });
  }
  return out;
}

function toIssues(raw: readonly RawIssue[]): ConfigIssue[] {
  return raw.map((issue) => ({ code: issue.code, path: pointer(issue.segments), message: issue.message }));
}

function loadSchema(value: unknown): Record<string, Json> {
  const json = toJson(value);
  if (!isJsonObject(json)) throw new Error("goondan.schema.json must be a JSON object");
  return json;
}

/** The packaged copy of `spec/goondan.schema.json`. */
export const configSchema: Record<string, Json> = loadSchema(schemaDocument);

/** Validates a value against the whole configuration schema. */
export function validateSchema(value: unknown, at: readonly PointerSegment[] = []): ConfigIssue[] {
  return sortIssues(toIssues(check(configSchema, configSchema, value, [...at])));
}

/** Validates a value against one `$defs` entry, reporting at `at`. */
export function validateDefinition(name: string, value: unknown, at: readonly PointerSegment[]): ConfigIssue[] {
  return toIssues(check(configSchema, resolveRef(configSchema, `#/$defs/${name}`), value, [...at]));
}

/** Validates a value against one root property subschema, reporting at `at`. */
export function validateRootProperty(name: string, value: unknown, at: readonly PointerSegment[]): ConfigIssue[] {
  const properties = configSchema.properties;
  if (!isRecord(properties)) throw new Error("The configuration schema has no properties");
  return toIssues(check(configSchema, asSchema(properties[name]), value, [...at]));
}

/** 호스트 도구가 선언한 JSON Schema로 호출 인수를 검사합니다. */
export function validateJsonValue(schema: Record<string, Json>, value: unknown): ConfigIssue[] {
  return toIssues(check(schema, schema, value, []));
}

/** Lists every schema location that uses a keyword outside the supported set. */
export function unsupportedSchemaKeywords(schema: unknown, at = "#"): string[] {
  if (typeof schema === "boolean") return [];
  if (!isRecord(schema)) return [`${at} is not a schema`];
  const problems: string[] = [];
  for (const [key, value] of Object.entries(schema)) {
    if (!supportedSchemaKeywords.has(key)) problems.push(`${at}/${key}`);
    else if (schemaMapKeywords.has(key) && isRecord(value)) {
      for (const [name, child] of Object.entries(value)) problems.push(...unsupportedSchemaKeywords(child, `${at}/${key}/${name}`));
    } else if (schemaKeywords.has(key)) problems.push(...unsupportedSchemaKeywords(value, `${at}/${key}`));
    else if (schemaListKeywords.has(key) && Array.isArray(value)) {
      value.forEach((child, index) => problems.push(...unsupportedSchemaKeywords(child, `${at}/${key}/${String(index)}`)));
    } else if (key === "$ref" && (typeof value !== "string" || !value.startsWith("#/"))) problems.push(`${at}/$ref is not local`);
  }
  return problems;
}
